src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
     self->visitMemory(&module->memory);
     self->visitModule(module);
   }
-};
 
-template<typename ParentType>
-struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>> {
-  ParentType& parent;
+  // Walk implementation. We don't use recursion as ASTs may be highly
+  // nested.
 
-  ChildWalker(ParentType& parent) : parent(parent) {}
+  // Tasks receive the this pointer and a pointer to the pointer to operate on
+  typedef void (*TaskFunc)(SubType*, Expression**);
```

Comment:
std::function is nicer in a way, I agree, but also more generic and hard to optimize (I doubt any compiler would reduce it to a single pointer).

Returning an expression has the downside that the visit* methods would need to return a value. Many (most?) passes don't need replacing, so it's extra annoying boilerplate, I think.